common/build/build-common/README.md

```diff
@@ -26,11 +26,16 @@ multiple config
 files](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#supporting-multiple-configuration-files-in-extends).
 
 -   tsconfig.base.json - This base config contains defaults that all packages within the repo should use as a baseline.
--   tsconfig.cjs.json - This config sets `module: Node16` and is intended for CommonJS builds. This config is intended to
-    be layered on top of the base config.
--   tsconfig.esm.json - This config sets `module: ESNext` and `moduleResolution: Node10` and is intended for ESM builds. This config
-    intended to be layered on top of the base config. This config uses Node10 module resolution because Node16 cannot be
-    used to build both CJS and ESM from a common source.
+-   tsconfig.cjs.json - This config sets `module: Node16` and `moduleResolution: Node16` and is intended for CommonJS
+    builds. This config is intended to be layered on top of the base config.
+-   tsconfig.esm.json - This config sets `module: ESNext` and `moduleResolution: Node10` and is intended for ESM builds
+    in packages that build both CJS and ESM. This config is intended to be layered on top of the base config. This config
+    uses `Node10` module resolution because `Node16` cannot be used to build both CJS and ESM from a common source.
```

Comment:
[This](https://stackoverflow.com/a/71473145) makes me a bit concerned (emphasis mine):

> when you say you’re using --module esnext, you’re allowed to write, and we will emit, the latest and greatest ES module code constructs, but we will not do anything differently with deciding how imports resolve. You’ll likely continue using --moduleResolution node, which was designed for Node’s implementation of CJS. What does this mean for you? **If you’re writing ESM for Node, you can probably make some stuff work with --module esnext and --moduleResolution node, but newer Node-specific features like [package.json exports](https://nodejs.org/api/packages.html#exports) won’t work**, and it will be extremely easy to shoot yourself in the foot when writing import paths. Paths will be evaluated by tsc under Node’s CJS rules, but then at runtime, Node will evaluate them under its ESM rules since you’re emitting ESM. There are significant differences between these algorithms—notably, the latter requires relative imports to use file extensions instead of dropping the .js, and index files have no special meaning, so you can’t import the index file just by naming the path to the directory.

Reply:
I think the way I describe it in the readme (this change) is misleading. TypeScript cannot produce ESM and CJS from the same source when the module resolution is Node16. The "problem" is that Node determines the module format based on file extension OR the nearest package.json's type field. tsc does not change file extensions or rewrite imports, so it alone cannot be used to dual-emit. Someone has to change the file extensions, because Node will _always_ pick a default module format - and it'll be CJS by default. Anyway, rather than write more here, I explain these problems and how we are addressing them in more detail in the release plan video, which I definitely recommend watching if you missed the meeting.

There were some early PRs that updated tsconfigs to use "invalid" combinations, but now that we're using tsc-multi we'll go back and unify everything. tsc-multi rewrites the imports and file extensions, so Node knows the module format for each file individually.
 
This PR is a bit unique since it's an ESM-only package, so we don't need tsc-multi, and we can use the .js file extension safely. We set type: module in the package.json, use the Node16 module type and moduleResolution (which are required to be together in newer versions of TS).

We also have a few CJS-only packages in the repo where we have done a similar thing and used tsc only but set type: commonjs explicitly in package.json.

Regarding the ts docs you noted, those problems are precisely why we use tsc-multi for dual-emit. Moreover, most of our packages have probably been broken in ESM Node for a while, since whenever this new module resolution behavior became the default in Node. We'll be in good shape once we've finished converting everything to Node16 moduleResolution (which implies tsc-multi for dual-emit packages).

Reply:
@RishhiB FYI.